hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -75,70 +90,118 @@ case class SegregatedBindingEnv[A, B](
     copy(newBindings = newBindings.bindRelational(bindings: _*))
 }
 
-case class EvalOnlyBindingEnv[T](env: Env[T]) extends GenericBindingEnv[EvalOnlyBindingEnv[T], T] {
-  override def promoteAgg: EvalOnlyBindingEnv[T] =
-    EvalOnlyBindingEnv(Env.empty)
-
-  override def promoteScan: EvalOnlyBindingEnv[T] =
-    EvalOnlyBindingEnv(Env.empty)
-
-  override def bindEval(bindings: (String, T)*): EvalOnlyBindingEnv[T] =
-    EvalOnlyBindingEnv(env.bindIterable(bindings))
-
-  override def dropEval: EvalOnlyBindingEnv[T] =
-    EvalOnlyBindingEnv(Env.empty)
-
-  override def bindAgg(bindings: (String, T)*): EvalOnlyBindingEnv[T] =
-    this
-
-  override def bindScan(bindings: (String, T)*): EvalOnlyBindingEnv[T] =
-    this
-
-  override def createAgg: EvalOnlyBindingEnv[T] =
-    this
-
-  override def createScan: EvalOnlyBindingEnv[T] =
-    this
-
-  override def noAgg: EvalOnlyBindingEnv[T] =
-    this
-
-  override def noScan: EvalOnlyBindingEnv[T] =
-    this
-
-  override def onlyRelational(keepAggCapabilities: Boolean = false): EvalOnlyBindingEnv[T] =
-    EvalOnlyBindingEnv(Env.empty)
-
-  override def bindRelational(bindings: (String, T)*): EvalOnlyBindingEnv[T] =
-    this
-}
-
 object Binds {
   def apply(x: IR, v: String, i: Int): Boolean =
-    Bindings(x, i, EvalOnlyBindingEnv(Env.empty[Type])).env.contains(v)
+    Bindings.get(x, i, BindingEnv.empty[Type].createAgg.createScan).eval.contains(v)
 }
 
+final case class Bindings(
+  eval: Seq[(String, Type)] = Seq.empty,
+  agg: AggEnv[Type] = AggEnv.NoOp,
+  scan: AggEnv[Type] = AggEnv.NoOp,
+  relational: Seq[(String, Type)] = Seq.empty,
+  dropEval: Boolean = false,
+)
+
 object Bindings {
+  val empty: Bindings = Bindings(Seq.empty, AggEnv.NoOp, AggEnv.NoOp, Seq.empty, false)
+
+  def dropAll(
```

Comment:
What does this `drop`? Doesn't it just construct a new instance?

Reply:
Ah this came from my misunderstanding of the default arguments!
I think what might be useful is to define `case class Bindings` with no default arguments and add "smart constructors" like `dropAll` that supply defaults that better describe the operation. I don't have a suggestion for what the default cause should be called, just please not "default" lol

Reply:
This is for when a node drops all bindings in the current environment, so that its descendents may not reference anything outside this node. This is used for all relational nodes.

Both this and the apply method are used constantly in this file, and I'd rather not make them more verbose. Maybe a comment describing what the defaults in `dropAll` mean, and what they're used for would help?

Note also that the following pr #14495 makes `dropAll` private.

Reply:
I prefer methods to describe what they do rather than have this is comments. I also don't think renaming `apply` to something else short like akin to `dropAll` will make things more verbose - you can always import that method once at the top and use it.

Since `dropAll` is being made private, i wont fight too much.